[VFABI] Improve VFABI unit tests

paschalis-mpeis opened the pull request:
Do checks for scalar parameter counts for all mappings with `matchScalarParametersNum`. It is not part of `invokeParser` so it can be selectively applied only to tests that supply valid mangled names.

Also, minor reformatting and typo fixes.

paschalis-mpeis committed:
[VFABI] Improve VFABI unit tests

Do checks for scalar parameter counts for all mappings with
`matchScalarParametersNum`. It is not part of `invokeParser` so it can
be selectively applied only to tests that supply valid mangled names.

Also, minor reformatting and typo fixes.

mgabka commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -44,10 +46,11 @@ class VFABIParserTest : public ::testing::Test {
   ElementCount &VF = Info.Shape.VF;
   VFISAKind &ISA = Info.ISA;
   SmallVector<VFParameter, 8> &Parameters = Info.Shape.Parameters;
+  size_t ScalarParametersNum;
```
i think would be good to initialise it to 0 here

paschalis-mpeis replied:
I didn't as there's a [reset](https://github.com/llvm/llvm-project/pull/73907/files/0ad4a31a051930353b1168b18216ff570b43f5c3#diff-c28f41966e76f1c7cf2a1b7f1e77e486f2e9f91fa7731fa0b7fd1cbead6f84d3R38)  method that deals with all other uninitialized variables when `invokeParser` gets called. I could move it closer to those variables (to line [46](https://github.com/llvm/llvm-project/pull/73907/files/0ad4a31a051930353b1168b18216ff570b43f5c3#diff-c28f41966e76f1c7cf2a1b7f1e77e486f2e9f91fa7731fa0b7fd1cbead6f84d3R46)) but if you think it's still not clear I could just initialize it.

mgabka commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -44,10 +46,11 @@ class VFABIParserTest : public ::testing::Test {
+  // References to the parser output field.
   ElementCount &VF = Info.Shape.VF;
   VFISAKind &ISA = Info.ISA;
   SmallVector<VFParameter, 8> &Parameters = Info.Shape.Parameters;
```
wouldn't be more clear to call VecFuncParamaters? to avoid confusion in the checks like:

  EXPECT_EQ(Parameters.size(), (unsigned)5);
  EXPECT_TRUE(matchScalarParametersNum());

where it is not clear why we check it twice.
WDYT?

mgabka commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -140,8 +147,10 @@ TEST_F(VFABIParserTest, OnlyValidNames) {
 
 TEST_F(VFABIParserTest, ParamListParsing) {
-  EXPECT_TRUE(invokeParser("_ZGVnN2vl16Ls32R3l_foo"));
+  EXPECT_TRUE(invokeParser("_ZGVnN2vl16Ls32R3l_foo", "",
```
so I think that this test should fail, as the mangled name is for a scalar function called foo, so I think we should faill if the scalar function name does not match.

paschalis-mpeis replied:
Added vector names to each test, and deriving the scalar one from the mangled name.
There should be tests for each of the names on each function that has a valid parsing.

paschalis-mpeis committed:
Further cleanup and simplication of VFABI tests.

`tryDemangleForVFABI` and `VFShape`'s `getScalarShape`, `get` use
`FunctionType` instead of `CallInst`.

Each successful parsing should provide a vector name, and do checks for
the names, number of arguments, and vector parameter types, eg:

```
EXPECT_EQ(ScalarName, "foo");
EXPECT_EQ(VectorName, "vector_foo");
...
EXPECT_TRUE(matchScalarParamNum());
...
EXPECT_EQ(VecFuncParameters[0], VFParameter({0, VFParamKind::OMP_Linear, 2}));
```

Replace method names to `foo` and `vector_foo` for the scalar and vector
variants respectively.
Using only `i32` for parameters when it's not relevant, except when
the VF Elements are explicitly checked, `ptr` for references, and `void`
for most return types.

All `VFABIParserTest` tests are now listed contiguously in the source.

huntergr-arm commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -34,20 +35,22 @@ class VFABIParserTest : public ::testing::Test {
   SMDiagnostic Err;
   std::unique_ptr<Module> M;
-  FunctionType *FTy;
+  FunctionType *FTy = nullptr;
```
Probably a good idea to change the name to ScalarFTy or similar to clarify.

paschalis-mpeis commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -131,209 +131,264 @@ TEST_F(VFABIParserTest, OnlyValidNames) {
-  EXPECT_EQ(VectorName, "_ZGVnM2v_sin");
+  EXPECT_TRUE(invokeParser("_ZGVnM2v_foo(vector_foo)"));
+  EXPECT_EQ(ScalarName, "foo");
+  EXPECT_EQ(VectorName, "vector_foo");
```
A separate patch will deal with the second commend, where `invokeParser` would reject such cases.
Essentially performing `matchScalarParamNum` in the parser.

paschalis-mpeis committed:
Addressing 2nd review round.

- Clean up duplicate ISA tests.
- Added an extra test to clearly show that the mangled name becomes the
  VectorName, when no VectorName is specified.

paschalis-mpeis committed:
Clang-format issue.

mgabka commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -103,6 +94,12 @@ class VFABIParserTest : public ::testing::Test {
   }
+
+  // Checks that the number of vectorized parameters matches the scalar ones.
+  // Takes into account that vectorized calls may also have a Mask.
```
could be worth to add that it requires correct ScalarFunction type to be used in the "invokeParser"

mgabka reviewed (commented):
I didn't review yet all the test changes.

paschalis-mpeis committed:
Made use for Scalar FunctionTypes more clear.

Removed ScalarFuncParametersNum, and now using the ScalarFTy directly,
that is parsed from the input ScalarFTyStr.

paschalis-mpeis committed:
Fix for demangle-fuzzer tool and minor refactoring

demangle-fuzzer had an extra parenthesis introduced by previous commit
in this PR and it was not formatted properly.

In ABI tests, renamed `reset` parameter to match `invokeParser`, and
did a few more minor styling changes.

mgabka commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -40,61 +38,45 @@ class VFABIParserTest : public ::testing::Test {
-  // generic fixed-length case can use as signature `void()`.
-  //
+
+  /// Invoke the parser. We need to make sure that a function exist in the
+  /// module because the parser fails if such function don't exists. Every time
```
I think that is not true anymore, as we moved away from computing scalable VF from the function declaration in the module, look at https://github.com/llvm/llvm-project/pull/72260

mgabka commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -40,61 +38,45 @@ class VFABIParserTest : public ::testing::Test {
+  ///
+  /// \p ScalarFTyStr FunctionType string to be used to get the signature of
+  /// the Scalar function. Used by `tryDemangleForVFABI` to check for the number
+  // of arguments on Scalable vectors, and by `matchScalarParamNum` to perform
```
misssing '/', also there is no need to use Scalable, all lower case is fine.

mgabka commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -40,61 +38,45 @@ class VFABIParserTest : public ::testing::Test {
-    // `invokeParser` multiple times in the same test.
-    reset(Name, IRType);
+                    const StringRef ScalarFTyStr = "void()") {
+    // Reset the VFInfo and the Module to be able to invoke `invokeParser`
```
mention of module can be removed

mgabka commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -40,61 +38,45 @@ class VFABIParserTest : public ::testing::Test {
     return false;
   }
 
   // Checks that 1. the last Parameter in the Shape is of type
```
I think that this function needs to be a bit more specific, i.e name changed to IsAArch64Masked() as looking at the https://software.intel.com/content/www/us/en/develop/download/vector-simd-function-abi.html specification for X86 the rules are different, i.e it says:
`"The total number of mask bits is equal to VLEN. The number of
mask parameters is equal to the number of parameters for the vector of characteristic data type"
....
"Mask parameters are passed after all other parameters in the same order of parameters that they apply to." `
If I am correct then some non aarch64 tests will need to be adjusted.

@huntergr-arm what is your understanding?


nit:If we want to keep this comment as enumeration, I think it would be nicer to refactor it as:

/// Checks if the function contains a vector mask compatible with AArch64 VFABI, i.e:
/// 1. The last parameter in the Shape is of type VFParamKind::GlobalPredicate.
/// 2. The function has only one  mask.

huntergr-arm replied:
I think that's a single mask which uses one or more vector registers depending on the datatypes and VF; if it takes more than one register to express, then it would be considered to be multiple parameters from a calling convention point of view even though its a single mask.

paschalis-mpeis committed:
Use VFInfo for `isMasked` and cleanup

Adjusted code and comments to address reviewers.
For masked calls, checking if the last parameter is a mask, only for
SVE and NEON ISAs.

Reverted VecFuncParameters to Parameters.

paschalis-mpeis commented on llvm/include/llvm/Analysis/VectorUtils.h:
```diff
@@ -96,22 +96,23 @@ struct VFShape {
   // with VF = 1.
-  static VFShape getScalarShape(const CallInst &CI) {
-    return VFShape::get(CI, ElementCount::getFixed(1),
+  static VFShape getScalarShape(const FunctionType *FTy) {
```
Opened new PR #74569

paschalis-mpeis commented on llvm/tools/vfabi-demangle-fuzzer/vfabi-demangler-fuzzer.cpp:
```diff
@@ -38,7 +38,8 @@ extern "C" int LLVMFuzzerTestOneInput(const uint8_t *Data, size_t Size) {
     std::unique_ptr<CallInst> CI(CallInst::Create(F, Args));
-    const auto Info = VFABI::tryDemangleForVFABI(MangledName, *(CI.get()));
+    const auto Info =
+        VFABI::tryDemangleForVFABI(MangledName, CI->getFunctionType());
```
Addressed in new PR #74569

paulwalker-arm commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -40,69 +50,48 @@ class VFABIParserTest : public ::testing::Test {
+  /// Returns whether the parsed function contains a mask
+  bool isMasked() const { return Info.isMasked(); }
+
+  /// Checks that the number of vectorized parameters matches the
```
"Check the number..."

paulwalker-arm commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -40,69 +50,48 @@ class VFABIParserTest : public ::testing::Test {
+
+  /// Checks that the number of vectorized parameters matches the
+  /// scalar ones. This requires a correct scalar FunctionType string to be fed
+  /// to the 'invokeParser'. It takes into account that vectorized calls may
+  /// also have a Mask.
```
Perhaps "Mask parameters that are only required by the vector function call are ignored.".

paulwalker-arm commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -423,107 +434,159 @@ TEST_F(VFABIParserTest, MissingVectorNameTermination) {
   EXPECT_EQ(Parameters[0], VFParameter({0, VFParamKind::Vector}));
-  EXPECT_EQ(Parameters[1], VFParameter({1, VFParamKind::GlobalPredicate}));
-  EXPECT_EQ(ScalarName, "sin");
+  EXPECT_EQ(ScalarName, "foo");
```
Is it deliberate to drop the check for `GlobalPredicate` from this and the following Masking tests?

mgabka replied:
@paulwalker-arm i think it is related to the comment on the top of this file:
3. The number of vector parameters and their types match the values
///    specified in the test.
///    On masked functions, when the VFISAKind is SVE or AdvancedSIMD (NEON) it
///    also checks that the last parameter is a mask (ie, GlobalPredicate).
///    TODO: Other VFISAKinds must add such checks according to their ABI
///    specification.

and looks reasonable to me.

mgabka replied:
ok, fair enough, won't argue about it.

paulwalker-arm commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -537,27 +600,40 @@ TEST_F(VFABIParserTest, ParseScalableMaskingSVESincos) {
 // Make sure we handle void return types.
 TEST_F(VFABIParserTest, ParseVoidReturnTypeSVE) {
-  EXPECT_TRUE(invokeParser("_ZGVsMxv_foo(vector_foo)", "foo", "void(i16)"));
+  EXPECT_TRUE(invokeParser("_ZGVsMxv_foo(vector_foo)", "void(i16)"));
+  EXPECT_TRUE(matchScalarParamNum()) << "Different number of scalar parameters";
```
Is there a test for a return but no parameters?

paschalis-mpeis replied:
Return a vector with no parameters? or `void()` ?

paulwalker-arm replied:
The former e.g. `float foo()` -> `<vscale x 4 x float> vector_foo()`.

paschalis-mpeis replied:
The grammar on LLVM [docs](https://llvm.org/docs/LangRef.html#call-site-attributes) for `<parameter>` isn't very clear, however for [AArch64](https://github.com/ARM-software/abi-aa/blob/main/vfabia64/vfabia64.rst#name-mangling-function) it is, and it looks like it requires 1+ parameters.

So I guess this is resolved?

paulwalker-arm replied:
It seems like a weird hole in the specification to me but as you say it is there in black and white so I guess so.  Thanks for investigating.

paschalis-mpeis committed:
Restored checks for predicates.

Checks in tests are now added in the same order:
1. Verify ISA
2. Verify presence/absence of a mask
3. match the number of parameters

Appended the tests if any of those were missing.

paschalis-mpeis committed:
Updated doxygen

mgabka commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -14,24 +15,31 @@
+/// Perform tests against VFABI Rules. `invokeParser` creates a VFInfo object
+/// and a scalar FunctionType, which are used by tests to check that:
+/// 1. The scalar and vector names are correct.
+/// 2. The parameter number from the parsed mangled name matches the scalar ones
```
nit: "The number of parameters from the parsed mangled name matches the number of arguments in the scalar function passed as FunctionType string."

mgabka commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -14,24 +15,31 @@
+///    specified in the test.
+///    On masked functions it also checks that the last parameter is a mask (ie,
+///    GlobalPredicate).
+/// 4. The function is correctly found to have a mask.
```
nit: The vector function

(since sometimes we mention here scalar functions as well)

mgabka commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -40,69 +48,48 @@ class VFABIParserTest : public ::testing::Test {
+  /// \p MangledName string the parser has to demangle.
+  ///
+  /// \p ScalarFTyStr FunctionType string to be used to get the signature of
+  /// the scalar function. Used by `tryDemangleForVFABI` to check for the number
```
nit: which is use by...

other wise the subject of the 2nd sentence is incorrect I think.

mgabka commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -40,69 +48,48 @@ class VFABIParserTest : public ::testing::Test {
-                      });
-    return NGlobalPreds == 1 && Info.Shape.Parameters.back().ParamKind ==
-                                    VFParamKind::GlobalPredicate;
+  /// Returns whether the parsed function contains a mask
```
nit: please add a dot at the end of the sentence.

mgabka commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -40,69 +48,48 @@ class VFABIParserTest : public ::testing::Test {
+  /// Returns whether the parsed function contains a mask
+  bool isMasked() const { return Info.isMasked(); }
+
+  /// Check the number of vectorized parameters matches the scalar ones. This
```
nit: "Check if the number..."

mgabka commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -423,108 +447,167 @@ TEST_F(VFABIParserTest, MissingVectorNameTermination) {
 }
```
definitely not for this patch, but at some point would be good to group somehow those tests in this file :)

paschalis-mpeis committed:
Improved comments.

paschalis-mpeis commented on llvm/unittests/Analysis/VectorFunctionABITest.cpp:
```diff
@@ -61,40 +61,31 @@ class VFABIParserTest : public ::testing::Test {
+  // \p STy -> FunctionType string to be used for the signature of
   // the vector function.  The correct signature is needed by the
   // parser only for scalable functions. For the sake of testing, the
   // generic fixed-length case can use as signature `void()`.
```
#76855 addresses this.
